File: ca-reservation/core/src/main/java/io/sharing/server/core/reservation/domain/ReservationStatus.java
package io.sharing.server.core.reservation.domain;

import java.util.List;

/**
 * 호스트의 경우
 * PENDING -> APPROVED
 *         -> DISAPPROVED
 *
 * REQUEST_CANCEL -> CANCELED
 *
 * 게스트의 경우
 * APPROVED -> REQUEST_CANCEL
 */
public enum ReservationStatus {
    REJECTED(),

    CANCELED(),

    REQUEST_CANCEL(),

    APPROVED(),

    PENDING();

//    private List<ReservationStatus> nextStatuses;

    boolean canChangeTo(ReservationStatus status) {
        boolean result = true;
        switch (this) {
            case REQUEST_CANCEL : if (status == CANCELED) result = false;
                break;
            case APPROVED : if (status == REQUEST_CANCEL) result = false;
                break;
            case PENDING : if (status == APPROVED || status == REJECTED) result = false;
                break;
        }

        return result;
    }

}

File: core/src/main/kotlin/io/sharing/server/core/reservation/domain/ReservationStatus.kt
package io.sharing.server.core.reservation.domain

/**
 * 호스트의 경우
 * PENDING -> APPROVED
 *         -> DISAPPROVED
 *
 * REQUEST_CANCEL -> CANCELED
 *
 * 게스트의 경우
 * APPROVED -> REQUEST_CANCEL
 */
enum class ReservationStatus(vararg status: ReservationStatus) {
    /** 거절 */
    REJECTED(),

    /** 취소됨 */
    CANCELED(),

    /** 취소 요청 */
    REQUEST_CANCEL(CANCELED),

    /** 승인 */
    APPROVED(REQUEST_CANCEL),

    /** 승인 대기 */
    PENDING(APPROVED, REJECTED);

    private val nextStatuses = status.toList()

    fun canChangeTo(status: ReservationStatus): Boolean {
        return status in nextStatuses
    }
}
